lightweight-term-count-update.php

```diff
@@ -12,79 +12,219 @@
  * @package 		Lightweight_Term_Count_Update
  */
 
+/**
+ * Main plugin class.
+ */
 class LTCU_Plugin {
 
-	// post stati which should be counted in term post counting.  Normally at least publish
-	public $counted_stati = array( 'publish' );		
+	/**
+	 * Post statuses which should be counted in term post counting. By default
+	 * this is [ 'publish' ], but it can be altered via the
+	 * `ltcu_counted_statuses` filter.
+	 *
+	 * @var array
+	 */
+	public $counted_statuses = array( 'publish' );
+
+	/**
+	 * Store the terms that have been incremented or decremented to avoid
+	 * duplicated efforts.
+	 *
+	 * @var array
+	 */
+	public $counted_terms = array();
+
+	/**
+	 * Store the singleton instance.
+	 *
+	 * @var My_Singleton
+	 */
+	private static $instance;
+
+	/**
+	 * Build the object.
+	 *
+	 * @access private
+	 */
+	private function __construct() {
+		/* Don't do anything, needs to be initialized via instance() method */
+	}
+
+	/**
+	 * Get the Singleton instance.
 
-	public function __construct() {
-		add_action( 'init', array( $this, 'init' ), 10, 0 );
+	 * @return My_Singleton
+	 */
+	public static function instance() {
+		if ( ! isset( self::$instance ) ) {
+			self::$instance = new static;
+			self::$instance->setup();
+		}
+		return self::$instance;
 	}
 
-	public function init() {
+	/**
+	 * Setup the singleton.
+	 */
+	public function setup() {
+		// Prevent core from counting terms.
+		wp_defer_term_counting( true );
 		remove_action( 'transition_post_status', '_update_term_count_on_transition_post_status' );
-		add_action( 'transition_post_status', array( $this, 'quick_update_terms_count' ), 10, 3 );
-		$this->counted_stati = apply_filters( 'LTCU_counted_stati',  $this->counted_stati );
+
+		$this->counted_statuses = apply_filters( 'ltcu_counted_statuses', $this->counted_statuses );
+		add_action( 'transition_post_status', array( $this, 'transition_post_status' ), 10, 3 );
+		add_action( 'added_term_relationship', array( $this, 'added_term_relationship' ), 10, 3 );
+		add_action( 'deleted_term_relationships', array( $this, 'deleted_term_relationships' ), 10, 3 );
 	}
 
-	public function quick_update_terms_count( $new, $old, $post ) {
-		global $wpdb;
+	/**
+	 * When a term relationship is added, increment the term count.
+	 *
+	 * @param int    $object_id Object ID.
+	 * @param int    $tt_id     Single term taxonomy ID.
+	 * @param string $taxonomy  Taxonomy slug.
+	 */
+	public function added_term_relationship( $object_id, $tt_id, $taxonomy ) {
+		$post = get_post( $object_id );
+		if ( in_array( $post->post_status, $this->counted_statuses, true ) ) {
```

Comment:
According to Justin's blog post, it might be moot to try to implement the enhancement for non-post objects:

> **`update_count_callback`:** You must use a custom function for this because WordPress expects the taxonomy to be for the post object type.

I'll dig into the code to see if this is still relevant, since that blog post is 5+ years old. I'll wait for your failing tests before writing any code regardless.